### PR TITLE
Add option for running ffmpeg_microphone_live as a background process

### LINKED ISSUE #30061: ffmpeg_microphone_live hangs if anything else is using STDIN
### System Info

- `transformers` version: 4.39.3
- Platform: Windows-10-10.0.19045-SP0
- Python version: 3.11.7
- Huggingface_hub version: 0.22.2
- Safetensors version: 0.4.2
- Accelerate version: not installed
- Accelerate config: not found
- PyTorch version (GPU?): 2.2.2+cpu (False)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: no
- Using distributed or parallel set-up in script?: no

When running ffmpeg_microphone_live in an environment where anything else is using STDIN, ffmpeg_microphone_live stops processing and prints `Enter command: <target>|all <time>|-1 <command>[ <argument>]`.

I believe this is related to this issue:
https://stackoverflow.com/questions/12543628/bash-script-stops-execution-of-ffmpeg-in-while-loop-why?rq=4

ffmpeg_microphone_live

### Who can help?

@sanchit-gandhi 

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

1. Run a process that uses stdin
2. Run a Python script that uses ffmpeg_microphone_live
3. Observe the behavior from the screenshot of outputting: `Enter command: <target>|all <time>|-1 <command>[ <argument>]`

### Expected behavior

I would hope that ffmpeg_microphone_live would have an option to use a different file for live reading or a -nostdin option with another file to support running as a background process

### PR DESCRIPTION
**Pull request by mikamerath**
# What does this PR do?
Gives the options for developers using audio_utils to fun `ffmpeg_microphone_live` as a background process instead of defaulting to STDIN which is the default option for ffmpeg. Useful for cases when running a different process that is using STDIN. The default case behavior is the same and now `ffmpeg_microphone_live` takes a boolean argument that allows the ffmpeg equivalent of passing in `-nostdin`.

Fixes #30061 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

- pipelines: @Narsil

### CODE REVIEW

**Comment by amyeroberts** [comment]
cc @ylacombe

**Comment by mikamerath** [comment]
@amyeroberts @ylacombe The project I'm working on is depending on this change. If there's any clarifications I can make about the motivation of this feature please let me know 😄

File: src/transformers/pipelines/audio_utils.py
Comment (amyeroberts) [inline src/transformers/pipelines/audio_utils.py:103]: There's no need to convert to string here

File: src/transformers/pipelines/audio_utils.py
Comment (amyeroberts) [inline src/transformers/pipelines/audio_utils.py:76]: Is this right? Shouldn't the default be `-stdin`?

File: src/transformers/pipelines/audio_utils.py
Comment (amyeroberts) [inline src/transformers/pipelines/audio_utils.py:168]:
```suggestion
            The boolean to determine whether or not to run the ffmpeg command in the background. If set to `True`,
            ffmpeg will not use STDIN as its input (useful for when other processes are using STDIN). If unset,
            ffmpeg will run using STDIN.
```

File: src/transformers/pipelines/audio_utils.py
Comment (amyeroberts) [inline src/transformers/pipelines/audio_utils.py:198]: no need for big if/else here 

```suggestion
    microphone = ffmpeg_microphone(
        sampling_rate,
        chunk_s,
        format_for_conversion=format_for_conversion,
        ffmpeg_input_device=ffmpeg_input_device,
        background_arg="-nostdin" if not run_in_background else "",
    )
```

File: src/transformers/pipelines/audio_utils.py
Comment (amyeroberts) [inline src/transformers/pipelines/audio_utils.py:74]: "background_arg" isn't very descriptive. It would be better to just propogate `run_in_background` to this function too if there's only two possible options and then set the correct string within this function

**Review by amyeroberts — COMMENTED** [review]
Hi @mikamerath, thanks for opening a PR! 

I've left some comments on the diff of this PR. The PR will need an approval from @ylacombe to see if it's something that the audio team would want to add. 

Requirement for a personal project isn't sufficient enough reason for adding something into the library. The great thing about open source is you can build upon and adapt the code to your own needs! The PR would be better motivated if you could provide a working example (code snippet) to demonstrate where this is useful and would be useful for other users

File: src/transformers/pipelines/audio_utils.py
Comment (ylacombe) [inline src/transformers/pipelines/audio_utils.py:181]: I'm not sure to follow here. If `run_in_background=False`, the microphone device is never defined? I'm pretty sure that it'd raise an error since it's used in the subsequent part of the code.

**Review by ylacombe — COMMENTED** [review]
Hey @mikamerath, thanks for opening up this PR!

As @amyeroberts suggested, could you explain in a few words why this PR is necessary? As I see it, the changes you make are not used anywhere else in the code base.

I've left a tiny question, besides Amy's remarks. I'll give a proper review once we agree on the next steps!

cc @eustlb if you want to give your opinion!

**Comment by mikamerath** [comment]
@amyeroberts @ylacombe thank you for looking at this. It is hard for me to get a minimal working example because the problem is a little bit convoluted. Today or tomorrow, I can attach a GIF showing the problem and I'll try to come up with a minimal script that reproduces the error.

**Comment by mikamerath** [comment]
NotWorking
Working


@amyeroberts @ylacombe Sorry for the long wait. This bug is hard to reproduce with a minimal example so I'll try to explain the motivation and what is causing it. We're using Whisper's speech to text for a project that has a local client/server connection. This connection uses IPC (which in turn uses STDIN) to pass messages back and forth.

When running our project, we would notice that part of the logs would include the message `Enter command: <target>|all <time>|-1 <command>[ <argument>]`. While investigating the error, I stumbled upon [This StackExchange Post](https://unix.stackexchange.com/questions/36310/strange-errors-when-using-ffmpeg-in-a-loop) that discussed similar errors when using ffmpeg. One of the solutions mentioned that ffmpeg uses STDIN by default and passing `-nostdin` which disables interaction. When adding `-nostdin` to the end of the arguments in `ffmpeg_command` the problem goes away, as demonstrated by the pictures attached at the top of this message. The first photo shows the error as it occurs now, while the second photo shows the behavior when this change has been made.

So I wanted to make a minimal change that would allow this feature to solve this issue. The default behavior stays the same, but it adds the ability to pass `-nostdin` so ffmpeg is not in interactive mode. I have tested it thoroughly with the present changes and it works perfectly. If you have additional questions or concerns, I'm happy to answer them.

**Comment by mikamerath** [comment]
@amyeroberts @ylacombe Any updates on your thoughts?

**Comment by amyeroberts** [comment]
@mikamerath Thanks for providing this information! This seems like quite a niche use-case and not an issue that hasn't been brought up before. 

I'd propose one of two things: 
* Adapting the microphone logic such that any additional arguments can be passed to ffmpeg to configure its usage. This way, we don't have to add and handle a new flag for everything in ffmeg. 
* Don't merge this PR and reopen if other people in the community request this feature in the future

**Comment by mikamerath** [comment]
@amyeroberts Sounds good, I'll update this pull request this weekend to allow any additional arguments to be passed to ffmpeg.

**Comment by mikamerath** [comment]
@amyeroberts The code has been updated to allow additional arguments to be passed into `ffmpeg_microphone_live` by passing a list of additional arguments via `ffmpeg_additional_args`

File: src/transformers/pipelines/audio_utils.py
Comment (amyeroberts) [inline src/transformers/pipelines/audio_utils.py:58]: No mutables as default arguments please! 

The right way to do this is have this unset to `None` by default and then default to an empty list if unset 

```suggestion
    ffmpeg_additional_args: Optional[List[str]] = None,
```

then later on

```py
    ffmpeg_additional_args = [] if ffmpeg_additional_args is None else ffmpeg_additional_args
```

File: src/transformers/pipelines/audio_utils.py
Comment (amyeroberts) [inline src/transformers/pipelines/audio_utils.py:36]: nit - formatting change shouldn't be applied here 

```suggestion
        with subprocess.Popen(ffmpeg_command, stdin=subprocess.PIPE, stdout=subprocess.PIPE) as ffmpeg_process:
```

File: src/transformers/pipelines/audio_utils.py
Comment (amyeroberts) [inline src/transformers/pipelines/audio_utils.py:41]: nit - same here 

```suggestion
        raise ValueError("ffmpeg was not found but is required to load audio files from filename") from error
```

File: src/transformers/pipelines/audio_utils.py
Comment (amyeroberts) [inline src/transformers/pipelines/audio_utils.py:78]: better to specify what should type of objects can contained in the list 

```suggestion
        ffmpeg_additional_args (`List[str]`, *optional*):
```

File: src/transformers/pipelines/audio_utils.py
Comment (amyeroberts) [inline src/transformers/pipelines/audio_utils.py:80]: It would be good to show a specific example here to show the user how to pass in the arguments e.g. `["-nostdin"]`. How should the user pass in if the arguments are a flag then a series of options e.g. `ffmpeg -flag argument_1 argument_2`?

File: src/transformers/pipelines/audio_utils.py
Comment (amyeroberts) [inline src/transformers/pipelines/audio_utils.py:309]: nit 

```suggestion
        print("ffmpeg was not found. Please install it or make sure it is in your system PATH.")
```

File: src/transformers/pipelines/audio_utils.py
Comment (amyeroberts) [inline src/transformers/pipelines/audio_utils.py:300]: nit 

```suggestion
        ffmpeg_devices = subprocess.run(command, text=True, stderr=subprocess.PIPE, encoding="utf-8")
        microphone_lines = [line for line in ffmpeg_devices.stderr.splitlines() if "(audio)" in line]
```

File: src/transformers/pipelines/audio_utils.py
Comment (amyeroberts) [inline src/transformers/pipelines/audio_utils.py:285]:
```suggestion
        raise ValueError("ffmpeg was not found but is required to stream audio files from filename") from error
```

File: src/transformers/pipelines/audio_utils.py
Comment (amyeroberts) [inline src/transformers/pipelines/audio_utils.py:276]:
```suggestion
        with subprocess.Popen(ffmpeg_command, stdout=subprocess.PIPE, bufsize=bufsize) as ffmpeg_process:
```

File: src/transformers/pipelines/audio_utils.py
Comment (amyeroberts) [inline src/transformers/pipelines/audio_utils.py:233]:
```suggestion
def chunk_bytes_iter(iterator, chunk_len: int, stride: Tuple[int, int], stream: bool = False):
```

File: src/transformers/pipelines/audio_utils.py
Comment (amyeroberts) [inline src/transformers/pipelines/audio_utils.py:216]:
```suggestion
    for item in chunk_bytes_iter(microphone, chunk_len, stride=(stride_left, stride_right), stream=True):
```

File: src/transformers/pipelines/audio_utils.py
Comment (amyeroberts) [inline src/transformers/pipelines/audio_utils.py:202]:
```suggestion
        raise ValueError(f"Unhandled format `{format_for_conversion}`. Please use `s16le` or `f32le`")
```

File: src/transformers/pipelines/audio_utils.py
Comment (amyeroberts) [inline src/transformers/pipelines/audio_utils.py:155]: This shouldn't be removed

File: src/transformers/pipelines/audio_utils.py
Comment (amyeroberts) [inline src/transformers/pipelines/audio_utils.py:167]: Same comments here - there should be an example of how to pass in multiple arguments in the docstring 

```suggestion
        ffmpeg_additional_args (`List[str]`, *optional*):
```

File: src/transformers/pipelines/audio_utils.py
Comment (amyeroberts) [inline src/transformers/pipelines/audio_utils.py:144]: Same comment here - no mutable object defaults 

```suggestion
    ffmpeg_additional_args: Optional[List[str]] = None,
```

**Review by amyeroberts — COMMENTED** [review]
Thanks for adding! 

Overall structure I think is OK - mainly some nits here and there on removing unnecessary formatting changes, making sure docstrings are as useful as possible and removing default empty lists as function defaults

**Comment by mikamerath** [comment]
@amyeroberts I believe everything should be good now. Let me know what you think.

**Review by ArthurZucker — APPROVED** [review]
She is OOO for a while, thanks for your contribution! IMO it's missing a small test but otherwise LGTM! 
Can you add one? 🤗

**Comment by mikamerath** [comment]
@ArthurZucker Yes, I'm planning on adding a test for passing in arguments and not passing in arguments. Does that seem like it covers everything that you would want to see?

**Comment by mikamerath** [comment]
@ArthurZucker Tests have been added.

**Comment by mikamerath** [comment]
Is there anything else I need to do?

**Review by ArthurZucker — COMMENTED** [review]
Thanks, sorry for the delay, merging!

**Pull request merged**
